### DISCUSSION TITLE
Go Router Navigating Through Same Page Multiple Times

### DISCUSSION THREAD

**Discussion by LLYX** [rrousselGit/riverpod#3007] [Q&A] [answered]
When navigating with go_router, it navigates me to that page multiple times. I print the current page path on every page in the logs, and it keeps showing double. I implemented the bare minimum without any providers, just go_router and provider scope wrapper, but the issue still occurs. I tried all the approaches mentioned in [this previous discussion](https://github.com/rrousselGit/riverpod/discussions/1357), but none of them seemed to help solve my problem.

Sample Logs:
> [GoRouter] Using MaterialApp configuration
flutter: Going to /
flutter: Going to /
[GoRouter] getting location for name: "dashboard"
[GoRouter] going to /dashboard
flutter: Going to /dashboard
flutter: Going to /dashboard
[GoRouter] getting location for name: "surveyInfo", pathParameters: {surveyId: 1}
[GoRouter] pushing /dashboard/survey-info/1
flutter: Going to /dashboard/survey-info/1
flutter: Going to /dashboard
flutter: Going to /dashboard

Code for my main:
```
void main() {
  runApp(const ProviderScope(
    child: MyApp(),
  ));
}

class MyApp extends ConsumerWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final router = ref.watch(goRouterProvider);

    return MaterialApp.router(
      routerConfig: router,
      title: 'Deliverzler',
      debugShowCheckedModeBanner: false,
      color: Theme.of(context).colorScheme.primary,
    );
  }
}

class MyHomePage extends ConsumerStatefulWidget {
  const MyHomePage({super.key, required this.title});
  final String title;

  @override
  MyHomePageState createState() => MyHomePageState();
}

class MyHomePageState extends ConsumerState<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    debugPrint("Going to ${GoRouterState.of(context).uri.toString()}");
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text(widget.title),
      ),
      body: ElevatedButton(
          onPressed: () => context.goNamed(DashboardPage.routeName),
          child: const Text("test")),
    );
  }
}
```

My go_router routes:
```
List<GoRoute> routes = [
  GoRoute(
      name: AppPage.home.name,
      path: AppPage.home.path,
      builder: (context, state) => const MyHomePage(title: 'Test'),
      routes: [
        GoRoute(
          name: AppPage.dashboard.name,
          path: AppPage.dashboard.path,
          builder: (context, state) => const DashboardPage(),
          routes: [
            GoRoute(
              name: AppPage.surveyInfo.name,
              path: "${AppPage.surveyInfo.path}/:${RouteParams.surveyIdKey}",
              builder: (context, state) => SurveyInfoPage(state),
              routes: [],
            )
          ],
        ),
      ]),
];
```

Any advice on what to do here would be appreciated.

**Comment by moxiaov587** [comment]
In the case of using `GoRouter`, usually you only need to pay attention to the logs printed by `GoRouter`.

The output of `debugPrint` does not mean that the route was navigated multiple times, it only means that `build` was executed multiple times.

**Reply by LLYX** [reply]
Hmm ok, though it only seems to happen when my app is wrapped by ProviderScope, otherwise the prints are as expected. Do you think that might be a problem or is that just how it is? Not sure why the page needs to be built multiple times when wrapped by ProviderScope, but only seems to build once without it.

**Reply by moxiaov587** [reply, marked as answer]
I still think this has nothing to do with `ProviderScope`,

I tested your code in my app and when the code was corrected to

```dart
debugPrint('MyHomePageState build');
```

The rebuild won't happen, I found [this issue](https://github.com/flutter/flutter/issues/123570), you can get more information from it, I now think the cause of the rebuild should be `GoRouterState.of(context)`.
